architecture/2024-08-resource-types-registration.md

```diff
@@ -0,0 +1,641 @@
+- (out of scope): Tooling for resource type registration (CLI, Bicep). This work is tracked by another design document.
+- (non-goal): Provider non-user-facing mechanisms for resource type registration.
+- (out of scope): A complete multi-locational architecture for UCP. 
+
+### User scenarios (optional)
+
+#### Defining new types
+
+Suzy is a platform engineer at Contoso working on the internal developer platform. She defines a resource type for an internal service `ContosoBus` named `Contoso.Platform/contosoBuses`. 
+
+She can use tools like Bicep, GitOps, or the `rad` CLI to register the new resource types in any tenant of Radius.
+
+*Note: the exact set of tooling implementations are out of scope for this document.*
+
+#### Removing default resource types
+
+Suzy is a platform engineer at Contoso working on the internal developer platform. She wants to avoid confusion about the set of resource types supported by the platform.
+
+The application teams are Contoso are allowed to use `Applications.Datastores/redisCaches` and `Applications.Datastores/postgreSQLDatabases`. They can't use other databases like `Applications.Datastores/sqlServerDatabases`. Suzy uses the `rad` CLI to unregister other types in the `Applications.Datastores` namespace.
+
+*Note: the exact set of tooling implementations are out of scope for this document.*
+
+## User Experience (if applicable)
+
+See API design. Tooling and user-experiences are out of scope for this document.
+
+## Design
+
+### High Level Design
+
+The registration APIs allow a user to control the set of resource types known to UCP. Since UCP is responsible for routing, it needs to know the set of resource providers and resource types so that it can send them traffic. 
+
+While resource types are the primary concept for **consumers** of extensibility (eg: application developers using Radius), resource providers are a significant concept for users that **implement** extensibility. The API design proposed here includes resource providers as a concept, and supports important scenarios for operating and versioning resource providers over time, based on lessons-learned from ARM.
+
+For anyone skeptical about this, the concept of resource provider is already present in **every** resource id/URL. This is the meaning of the `providers` segment. Currently URLs like `/planes/radius/local/providers` don't have any functionality associated with them, and return a 404 - this proposal addresses this gap.
+
+One of these lessons-learned is the ability to register multiple regions/locations where a resource provider can operate. In particular users will need the ability to gradually roll out functionality across regions rather than updating all regions at once. This is complex, and doesn't help us in the short term. The approach taken here is aligned with our current strategy for region/location in Radius. The APIs are designed for multi-region but the implementation assumes a single region is used.
```

Comment:
Will they still have the option to roll out functionality to all regions it they want? Like passing in **/*** for the location parameter in the create query for example.

Reply:
I think you're on the right track. Exactly how users do that is out of scope for this doc.